llvm/include/llvm/Analysis/TargetTransformInfo.h

```diff
@@ -495,6 +495,13 @@ class TargetTransformInfo {
 
   LLVM_ABI bool isNoopAddrSpaceCast(unsigned FromAS, unsigned ToAS) const;
 
+  LLVM_ABI std::pair<KnownBits, KnownBits>
+  computeKnownBitsAddrSpaceCast(unsigned FromAS, unsigned ToAS,
+                                const Value &PtrOp) const;
+
+  LLVM_ABI KnownBits computeKnownBitsAddrSpaceCast(
```

Comment:
Why do these two functions have to be in TTI?

Reply:
The addrspacecast. For example, the case where I cast larger address to smaller address, I just trunc. This might be valid for all practical cases (given integral), but it's in no way guaranteed afaiu.
Also, the anyext for smaller->larger casts is a conservative choice that could be improved with target knowledge, ig

Reply:
I already replaced the zext by an anyext in a previous commit, so I don't assume anything specific about the casts afaics. Except that larger->smaller casts truncate. But that may be too much to assume, either.

Reply:
The best option would be to add a case for `Instruction::AddrSpaceCast` to `computeKnownBitsFromOperator` in `ValueTracking.cpp`. But since we cannot access TTI functions from there, ig that's not an option.
This means, however, that `computeKnownBitsAddrSpaceCast` should stay recursive. Otherwise, addrspacecasts as source values will never be handled correctly and I don't think the user of `computeKnownBitsAddrSpaceCast` should be required to build a loop around it?

Reply:
Ping. If there is not really a good solution, ig the best option would be to remove the target implementation of computeKnownBits for AddrSpaceCast that Matt suggested? So that we can at least fix the original bug our fuzzer detected.